src/algebraic_geometry/projective_spectrum/scheme.lean

```diff
@@ -120,36 +120,281 @@ def degree_zero_part {f : A} {m : ℕ} (f_deg : f ∈ 𝒜 m) : subring (away f)
+
+  rcases x.1.is_prime.mem_or_mem (show a1 * a2 * f ^ N * f ^ M ∈ _, from _) with h1|rid2,
+  rcases x.1.is_prime.mem_or_mem h1 with h1|rid1,
+  rcases x.1.is_prime.mem_or_mem h1 with h1|h2,
+  { left,
+    rw mem_carrier_iff,
+    simp only [show (mk a1 ⟨f ^ n1, _⟩ : away f) = mk a1 1 * mk 1 ⟨f^n1, ⟨n1, rfl⟩⟩,
+     by erw [localization.mk_mul, mul_one, one_mul]],
+    exact ideal.mul_mem_right _ _ (ideal.subset_span ⟨_, h1, rfl⟩), },
+  { right,
+    rw mem_carrier_iff,
+    simp only [show (mk a2 ⟨f ^ n2, _⟩ : away f) = mk a2 1 * mk 1 ⟨f^n2, ⟨n2, rfl⟩⟩,
+      by erw [localization.mk_mul, mul_one, one_mul]],
+    exact ideal.mul_mem_right _ _ (ideal.subset_span ⟨_, h2, rfl⟩), },
+  { exact false.elim (x.2 (x.1.is_prime.mem_of_pow_mem N rid1)), },
+  { exact false.elim (x.2 (x.1.is_prime.mem_of_pow_mem M rid2)), },
+  { rw eq1,
+    refine mul_mem_right _ _ (mul_mem_right _ _ (sum_mem _ (λ i hi, mul_mem_left _ _ _))),
+    generalize_proofs _ h,
+    exact (classical.some_spec h).1 },
+end⟩
+
+lemma preimage_eq (a : A) (n : ℕ)
+  (a_mem_degree_zero : (mk a ⟨f ^ n, ⟨n, rfl⟩⟩ : away f) ∈ A⁰_ f_deg) :
+  to_fun 𝒜 f_deg ⁻¹'
+      (sbo (⟨mk a ⟨f ^ n, ⟨_, rfl⟩⟩, a_mem_degree_zero⟩ : A⁰_ f_deg)).1
+  = {x | x.1 ∈ (pbo f) ⊓ (pbo a)} :=
+begin
+  haveI : decidable_eq (away f) := classical.dec_eq _,
+  ext1 y, split; intros hy,
+  { refine ⟨y.2, _⟩,
+    erw [set.mem_preimage, prime_spectrum.mem_basic_open] at hy,
+    erw projective_spectrum.mem_basic_open,
+    intro a_mem_y,
+    apply hy,
+    rw [to_fun, mem_carrier_iff],
+    simp only [show (mk a ⟨f^n, ⟨_, rfl⟩⟩ : away f) = mk 1 ⟨f^n, ⟨_, rfl⟩⟩ * mk a 1,
+      by erw [mk_mul, one_mul, mul_one]],
+    exact ideal.mul_mem_left _ _ (ideal.subset_span ⟨_, a_mem_y, rfl⟩), },
+  { change y.1 ∈ _ at hy,
+    rcases hy with ⟨hy1, hy2⟩,
+    erw projective_spectrum.mem_basic_open at hy1 hy2,
+    rw [set.mem_preimage, to_fun],
+    erw prime_spectrum.mem_basic_open,
+    intro rid,
+    rcases mem_carrier.clear_denominator f_deg _ rid with ⟨c, N, acd, eq1⟩,
+    rw [mk_mul, mul_one] at eq1,
+    simp only [mk_eq_mk', is_localization.eq] at eq1,
+    rcases eq1 with ⟨⟨_, ⟨M, rfl⟩⟩, eq1⟩,
+    erw [mul_one] at eq1,
+    simp only [← subtype.val_eq_coe] at eq1,
+
+    rcases y.1.is_prime.mem_or_mem (show a * f ^ N * f ^ M ∈ _, from _) with H1 | H3,
+    rcases y.1.is_prime.mem_or_mem H1 with H1 | H2,
+    { exact hy2 H1, },
+    { exact y.2 (y.1.is_prime.mem_of_pow_mem N H2), },
+    { exact y.2 (y.1.is_prime.mem_of_pow_mem M H3), },
+    { rw eq1,
+      refine mul_mem_right _ _ (mul_mem_right _ _ (sum_mem _ (λ i hi, mul_mem_left _ _ _))),
+      generalize_proofs _ h,
+      rcases classical.some_spec h with ⟨h1, -⟩,
+      exact h1, }, },
+end
+
+end forward
+
+section
+
+variable {𝒜}
+
+/--The continuous function between the basic open set `D(f)` in `Proj` to the corresponding basic
+open set in `Spec A⁰_f`.
+-/
+def forward {f : A} (m : ℕ) (f_deg : f ∈ 𝒜 m) :
```

Comment:
I think this name is a bit ambiguous. Maybe `to_Spec` would be a better name?

Reply:
By ambiguous I meant that I couldn't really get what `Top_component.forward` / `backward` means without looking at the docstring or the type. IMO something like `Proj_Top_component.to_Spec` / `from_Spec` would be clearer?

Reply:
I changed to `Proj_iso_Spec_Top_Component.to_Spec`